api/src/main/java/org/pmiops/workbench/audit/adapters/WorkspaceAuditAdapterServiceImpl.java

```diff
@@ -73,23 +81,135 @@ private void logAndSwallow(RuntimeException e) {
   @Override
   public void fireDeleteAction(org.pmiops.workbench.db.model.Workspace dbWorkspace) {
     try {
-      final String actionId = ActionAuditService.newActionId();
+      final String actionId = ActionAuditEvent.newActionId();
       final long userId = userProvider.get().getUserId();
       final String userEmail = userProvider.get().getEmail();
-      final ActionAuditEvent event =
-          new ActionAuditEvent.Builder()
+      final long timestamp = clock.millis();
+      final ActionAuditEventImpl event =
+          new ActionAuditEventImpl.Builder()
               .setActionId(actionId)
               .setAgentEmail(userEmail)
               .setActionType(ActionType.DELETE)
               .setAgentType(AgentType.USER)
               .setAgentId(userId)
               .setTargetType(TargetType.WORKSPACE)
               .setTargetId(dbWorkspace.getWorkspaceId())
+              .setTimestamp(timestamp)
               .build();
-
       actionAuditService.send(event);
     } catch (RuntimeException e) {
       logAndSwallow(e);
     }
   }
+
+  @Override
+  public void fireDuplicateAction(
+      org.pmiops.workbench.db.model.Workspace sourceWorkspaceDbModel,
+      org.pmiops.workbench.db.model.Workspace destinationWorkspaceDbModel) {
+    try {
+      // We represent the duplication as a single action with events with different
+      // ActionTypes: DUPLICATE_FROM and DUPLICATE_TO. The latter action generates many events
+      // (similar to how CREATE is handled) for all the properties. I'm not (currently) filtering
```

Comment:
So... even though it's a single workspace, each auditable property set on the new workspace is its own event?